interaction-manager/src/lib_components/dropzone.js

```diff
@@ -0,0 +1,311 @@
+        const { x: elementX, y: elementY } = this.draggableElements[elementIndex].getBoundingClientRect();
+
+        const direction = {
+            x: Math.sign(x - elementX),
+            y: Math.sign(y - elementY),
+        };
+
+        this.draggableElements[elementIndex].style.position = 'absolute';
+
+        const loop = (newX, newY) => {
+            if (newX === x && newY === y) {
+                this.dropzones[dropzoneIndex].appendChild(this.draggableElements[elementIndex]);
+                this.draggableElements[elementIndex].style.position = '';
+                return;
+            }
+
+            newX = newX !== x ? direction.x + newX : newX;
+            newY = newY !== y ? direction.y + newY : newY;
+
+            this.draggableElements[elementIndex].style.left = `${newX}px`;
+            this.draggableElements[elementIndex].style.top = `${newY}px`;
+
+            if (newX !== x || newY !== y) requestAnimationFrame(() => loop(newX, newY));
+        };
+
+        loop(elementX, elementY);
+    }
+
+    /**
+     * Saves the dropzones
+     */
+    createDropzones() {
+        this.dropzones = this.options.dropzones.reduce((acc, dropzone) => {
+            const dropzoneElements = [...document.querySelectorAll(dropzone)];
+            if (dropzoneElements.length === 0) {
+                console.error(`${dropzone} is not a valid html element to be used as a dropzone`);
+            }
+
+            return acc.concat(dropzoneElements);
+        }, []);
+    }
+
+    /**
+     * Handler when you drop the dragged item
+     */
+    handleDrop() {
+        let dropType = this.options.dropType;
+
+        if (this.draggedOver.children.length === 0) dropType = 'add';
+
+        switch (dropType) {
+            case 'add':
+                this.draggedOver.appendChild(this.draggedElement);
+                break;
+            case 'shift':
+                this.shiftElements();
+                break;
+            case 'switch':
+                this.draggedElement.parentNode.appendChild(this.draggedOver.children[0]);
+                this.draggedOver.appendChild(this.draggedElement);
+                break;
+            case 'none':
+                break;
+            default:
+                break;
+        }
+
+        this.onMouseLeave();
+    }
+
+    /**
+     * Shifts the element to the nearest empty space
+     */
+    shiftElements() {
+        const dropzoneIndex = this.dropzones.findIndex(dropzone => dropzone === this.draggedOver);
+        const dropzoneChildren = this.dropzones.map((dropzone) => {
+            if (dropzone.children[0] === this.draggedElement) return [];
+            return [...dropzone.children];
+        });
+
+        const closestEmptyIndex = dropzoneChildren.reduce((acc, el, index) => {
+            if (el.length === 0) acc = Math.abs(dropzoneIndex - index) < Math.abs(dropzoneIndex - acc) ? index : acc;
+            return acc;
+        }, 100000);

```

Comment:
Why such a specific number?

Reply:
In reduce, when you don't provide an initial value, it sets the first item of the array as the initial value, then starts the loop from the second. Since in this case the elements are HTMLElements and we are measuring distance, we need to set a big number so that it can pass the check. I can also use Inifinity if you think it's better readable.